### ISSUE TITLE
Missing timezone in metadata date can lead to wrong date in url

### ISSUE THREAD

**Issue by zwippie** [jekyll/jekyll#1069]
Starting with a fresh new Jekyll site with no modifications, I noticed that the first auto-generated post got the wrong page date when served. It looks like a missing timezone in the posts metadata can lead to the generated url being one day off when the time is close enough to midnight.

Metadata of file `/_posts/2013-05-09/welcome-to-jekyll.markdown`:

```

---
layout: post
title:  "Welcome to Jekyll!"
date:   2013-05-09 22:39:53
categories: jekyll update

---
```

This leads to the post having url `/jekyll/update/2013/05/10/welcome-to-jekyll.html`, which can be explained by me being in a timezone with offset +0200.

The url I expected to see was `/jekyll/update/2013/05/09/welcome-to-jekyll.html`.

Changing or removing the time part of the metadata (or adding timezone info) fixes the problem as expected, but it left me puzzled for a moment.

**Comment by alanguir** [comment]
I'm experiencing this as an issue too, and it definitely feels broken. I'd only expect the output of the date filter to match UTC if I were, in fact, on UTC +0000. I didn't notice this when working just with dates, but it was immediately apparent when trying to work with hours. 

IRB tells me I have my time set correctly
```
2.2.0 :001 > Time.now
 => 2015-03-06 11:57:35 -0500 
```

`_config.yml` has the timezone set correctly: 

```yaml
timezone: America/Detroit
markdown: kramdown
encoding: utf-8
...
```

Post front matter has a dates array in local time

```
dates:
  - start: 2015-3-23 18:00:00
    end: 2015-3-23 21:00:00
  - start: 2015-3-25 18:00:00
    end: 2015-3-25 21:00:00
```

No funky tricks in the date rendering:

```
{% for date in event.dates %}
  <li>{{ date.start | date: '%b %d, %Y  %l:%M %P' }} - {{ date.end | date: ' %l:%M %P' }}</li>
{% endfor %}
```

And the rendered HTML shows times in UTC. ( that should be 6pm (18:00) - 9pm(21:00), rather than 2pm and 5pm respectively ). 

```
Mar 23, 2015 2:00 pm - 5:00 pm
Mar 25, 2015 2:00 pm - 5:00 pm
```

Miraculously, jekyll does correctly calculate daylight savings time by only rending the UTC time as -0400 instead of the usual  -0500 for my part of the world. 

ps - jekyll 2.5.3

**Comment by alanguir** [comment]
Update: setting my timezone to `Africa/El_Aaiun` in `_config.yml` in the meantime to avoid adding the offset to each date in each post. [Apparently](http://en.wikipedia.org/wiki/List_of_tz_database_time_zones) that will give me the same DST rules as US Eastern time, and output my times as expected.

**Comment by parkr** [comment]
@alanguir This behaviour is per `safe_yaml` and `Time.parse` when you give us a string. We set the `TZ` variable before these methods are executed, but if you have no offset, Ruby has no way to know what that time should be so it just forgets about it and assigns it to UTC at all times (IIRC). Going back and adding offsets is highly recommended. It can be a pain, but it's important, otherwise you have over 24 times in one (one per timezone), so there's no way to know what exact moment in time you're looking for.

**Comment by alanguir** [comment]
@parkr I understand how this could cause unexpected results if I'm writing my times assuming UTC +5, and jekyll has a build script that runs when I deploy on a heroku server living in UTC +8, but why ignore the explicit timezone setting in the config file? That seems like a perfectly good way to know what the time should be. 

What is the timezone config option used for if not helping jekyll parse dates at build time?

**Comment by parkr** [comment]
> What is the timezone config option used for if not helping jekyll parse dates at build time?

All it does is set `$TZ` in your Ruby environment. If you have no timezone data, it doesn't add anything in (it doesn't default to your local timezone). It makes me sad. I [wrote a blog post about this sort of thing](https://byparker.com/blog/2014/ruby-2-2-0-time-parse-localtime-regression/) (though not this _exact_ thing). Timezones in Ruby aren't great.

**Comment by willnorris** [comment]
yep, I ran into the crappy support for timezones in Ruby when I switched to Jekyll too.  In my case, I very specifically want to preserve the timezone of when a post was written, and display that on my site (for example, compare by-lines [here](https://willnorris.com/2006/01/life-and-love-and-why) vs [here](https://willnorris.com/2015/02/go-get-subpackages-nginx)).  For me, that additional bit of data is meaningful.

I do this by ensuring that I always specify dates as strings in my frontmatter (otherwise the yaml module will "helpfully" parse them as `Time` objects), and then use a [custom jekyll plugin](https://github.com/willnorris/willnorris.com/blob/master/src/_plugins/datetime.rb) to parse the post date as a `DateTime`, which preserve timezone information.  Not sure if that is helpful to others in this thread, but just wanted to point out one way I've worked around ruby's TZ shortcomings.

**Comment by parkr** [comment]
@willnorris Do you see a bugfix path for safe_yaml or Jekyll in this regard?

**Comment by willnorris** [comment]
@parkr I've never actually looked closely at what safe_yaml does before, but now this does make sense: [date.rb](https://github.com/dtao/safe_yaml/blob/master/lib/safe_yaml/parse/date.rb).  It parses dates as `DateTime`, but then converts them to `Time` and always returns the value in local time.  In my opinion, this is totally the wrong thing to do since (if I'm correctly understanding how time is handled in ruby) you're throwing away any TZ data that was there.  I imagine having safe_yaml return the `DateTime` object would be a pretty major breaking change for them and not realistic.  The only thing Jekyll could do is basically what I'm doing... if a date is specified as a string, then you could additionally parse it as a `DateTime` and preserve the TZ info that way.

**Comment by parkr** [comment]
> It parses dates as DateTime, but then converts them to Time and always returns the value in local time. In my opinion, this is totally the wrong thing to do since (if I'm correctly understanding how time is handled in ruby) you're throwing away any TZ data that was there.

So they fixed this in Ruby 2.2.0, which broke Jekyll's expectations that it'd be in your local timezone (that's what [my blog post is about](https://byparker.com/blog/2014/ruby-2-2-0-time-parse-localtime-regression/)).

> I imagine having safe_yaml return the DateTime object would be a pretty major breaking change for them and not realistic.

@dtao, how would you feel about this?

> if a date is specified as a string, then you could additionally parse it as a DateTime and preserve the TZ info that way.

Right now we use `Time.parse` – should we switch to using `DateTime.parse`?

**Comment by willnorris** [comment]
> > It parses dates as DateTime, but then converts them to Time and always returns the value in local time. In my opinion, this is totally the wrong thing to do since (if I'm correctly understanding how time is handled in ruby) you're throwing away any TZ data that was there.

> So they fixed this in Ruby 2.2.0, which broke Jekyll's expectations that it'd be in your local timezone (that's what [my blog post is about](https://byparker.com/blog/2014/ruby-2-2-0-time-parse-localtime-regression/)).

Oh, interesting.. so I somehow missed that.  So now this means that ruby 2.2 is preserving the TZ information correctly, but now it's Jekyll that is throwing it away by always calling `localtime` :frowning:  I know that this is how Jekyll has always worked historically and perhaps has some assumptions built in for this, but it still makes me sad because we're throwing away the data.  If a post date is specified without TZ information, then sure treat it as localtime.  But if an author has taken the effort to specify the TZ in the post date, then it seems very unfortunate that we don't respect and preserve that.

> > if a date is specified as a string, then you could additionally parse it as a DateTime and preserve the TZ info that way.

> Right now we use `Time.parse` – should we switch to using `DateTime.parse`?

Maybe?  I'm not sure where else `Time` and `DateTime` differ and what problems this could potentially introduce.  I seem to recall that some of our date-related liquid filters choked on a `DateTime`, but don't remember fully.

**Comment by pathawks** [comment]
> But if an author has taken the effort to specify the TZ in the post date, then it seems very unfortunate that we don't respect and preserve that.

:+1: This has also often bothered me.

**Comment by alanguir** [comment]
To solve my specific issue, I wrote a tiny Jekyll filter that will apply your local timezone offset to a custom front matter date. Be careful when using with something like a heroku deploy script - I haven't tested there, although tzinfo is explicitly setting the timezone based on Jekyll config:

https://gist.github.com/alanguir/72c6d22d95f3ed80a511

**Comment by sumdog** [comment]
I just started using Jeykll and just noticed this. I have converted posts from a legacy/proprietary system where each post had a date and a time (but no timezone). Everything was a full day off (My localtime is set to Australia/Melbourne). 

I feel like if I don't specify a timezone, than those are the dates/times I want to use when displaying to the user. They really shouldn't go through any conversion at all. (and even if I do specify a timezone...there should really only be a conversion if the _config.yml has a different timezone).  We're talking about static content here; not content that changes depending on the connecting browser. 

There should be an option to disable any timezone alteration entirely, although judging from the comments, I'm not sure if that's entirely possibly if this is depending on the underlying YAML engine.

**Comment by envygeeks** [comment]
This is more than likely a problem with the object we use rather than the way we use the object, DateTime sucks when an explicit timezone isn't supplied and if I remember we are using DateTime objects?

**Comment by jaypinho** [comment]
Just ran into the same problem here as well.

Posted something at 1:22 AM New York time (UTC - 4 hours) and put in the data without time zone as '2015-05-22 01:22:00' (assuming it would show up correctly since I have 'timezone: America/New_York' in my _config.yml file), but the URL slug and the post date both show up as the day before. This doesn't seem like the type of behavior you'd expect by default.

**Comment by chemoish** [comment]
Trying to follow years of timezone posts,,, It would be nice to be able to specify a time and have the `timezone` modify that time appropriately on compilation.

`2017-02-12 14:30:00` + `America/Los_Angeles` = `2017-02-12 14:30:00 -0800`
`2017-04-02 14:30:00` + `America/Los_Angeles` = `2017-04-02 14:30:00 -0700`

/shrug

**Comment by scottdorman** [comment]
This issue seems to still exist. Look at ~~https://scottdorman.github.io/2015/05/29/introducing-jquery-gotop/~~ https://scottdorman.github.io/2015/05/28/introducing-jquery-gotop/as an example. The date portion of the link gets set to `2015/05/29` even though the front matter has `date: '2015-05-28 23:23:39.00000000 -05:00'` and my _config.yml has `timezone: America/New_York`. I have tried the front matter date in several variations and none of them seem to make any difference.

This issue affects a bunch of pages for me and also has an impact on some automatic redirection logic I have set up since the date portions of the links don't match up.

If there's something I need to change, I'm all for changing it but it seems like this is an issue with how Jekyll is parsing the date front matter variable to generate the URL. If the date has timezone information provided (which mine does) it should always use that timezone.

Interestingly enough, as I was writing this I decided to try setting the date to the corresponding UTC time (2015-05-29 03:23:39) which still had no effect. However, when I set the date back to the original and **removed** the time zone portion, it was generated with the correct date in the URL.  This behavior seems widly inconsistent, as other pages seem to work just fine with the time zone in the date (as a string) (see https://github.com/scottdorman/scottdorman.github.io/blob/master/_posts/2018/2018-08-18-saving-bootstrap-component-state.md as an example). It seems like where things start to go wrong is when the time approaches that roll-over period in UTC (so, for my time zone any dates earlier than 19:00:00 generate with the correct date and anything after would appear to generate with the wrong date).
